ismatec/util.py

```diff
@@ -220,59 +225,63 @@ def close(self):
 
 
 class Protocol:
```

Comment:
This is one of the bigger (but still nitpicky) areas I've been struggling with. On one hand, a protocol should be more than just bizarre datatype conversion. It should be everything about building requests and parsing responses.

I started a version of this that looked like...

```python
class Protocol:
    """Convert to various (dumb) datatypes used for the protocol."""
    requests = {
        'pump version': '1#',
        'protocol version': '1x!',
        'set flow rate': '{channel}f{flow_v2}',
        'get flow rate': '{channel}f',
        'get mode': '{channel}xM',
        'set mode': ...

    }
    responses = {
        'setpoint': {
            '0': 'rpm',
            '1': 'flow rate',
        },
        'rotation': {
            'J': 'clockwise',
            'K': 'counterclockwise',
        },
        'mode': {
            'G': 'vol over time',
            'L': 'rpm',
            'M': 'flow rate',
            'N': 'time',
            'O': 'vol at rate',
            'P': 'time pause',
            'Q': 'vol pause',
        },
    }
    tubing = [
        0.13, 0.19, 0.25, 0.38, 0.44, 0.51, 0.57, 0.64, 0.76, 0.89, 0.95, 1.02, 1.09,
        1.14, 1.22, 1.30, 1.43, 1.52, 1.65, 1.75, 1.85, 2.06, 2.29, 2.54, 2.79, 3.17
    ]
```

I stopped because it felt like an over-abstraction. Your driver is mostly <5 lines per function, so we're not gaining much readability in exchange for the complexity.

I'd recommend deleting `Protocol` altogether. Either tack the unit conversion to the bottom of `driver` or make them functions in `util`. Leaning the latter, where you'd call `util.pack_time1` in the driver.

Reply:
Muahah, mypy caught your breaking change.  (which I'm not opposed to, it just needs to be complete)

Reply:
What do you think about `StrEnum`? 


It feels like a good fit here. Added in 3.11, but there's a compatible PyPI package for older versions.

Reply:
Somewhat neutral on the choice, outside of thinking about what the mindless doesnt-care-about-implementation end-user experience would be.

It might be as simple as `print(pump.get_mode())` printing a reasonable output, which could be done by defining `__repr__` in the enum. 

Maybe people would also want a general `pump.get()` that they'd serialize into JSON and send over a wire? If that's the case, they'd want the output to be a serializable object (aka string).

Repeating myself - this is all in the interest of extreme nitpickiness and this driver is already better than many you'll find online.